File: 백준/Gold/2110. 공유기 설치/공유기 설치.c
#include <stdio.h>
#include <stdlib.h>

int cmp(const void* a, const void* b) {
	if (*(int*)a > *(int*)b) {
		return 1;
	}
	else if (*(int*)a < *(int*)b) {
		return -1;
	}
	else {
		return 0;
	}
}

int main(void) {
	int n, c, i, left, right, mid, prev, cnt;
	int* an;
	scanf("%d %d", &n, &c);
	an = (int*)malloc(n * sizeof(int));
	for (i = 0; i < n; i++) {
		scanf("%d", &an[i]);
	}
	qsort(an, n, sizeof(int), cmp);
	left = 1;
	right = an[n - 1] - an[0];
	while (left <= right) {
		mid = (left + right) / 2;
		prev = an[0];
		cnt = 1;
		for (i = 1; i < n; i++) {
			if (an[i] - prev >= mid) {
				cnt++;
				prev = an[i];
			}
		}
		if (cnt >= c) {
			left = mid + 1;
		}
		else {
			right = mid - 1;
		}
	}
	printf("%d", right);
	return 0;
}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        int n;
        int c;
        int i;
        int cnt;
        int mid;
        int prev;
        int left;
        int right;
        int[] locs;
        BufferedReader br;
        StringTokenizer st;

        br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine(), " ", false);
        n = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken()) - 1;
        locs = new int[n];
        for (i = 0; i < n; i++) {
            locs[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(locs);
        left = 1;
        right = locs[n - 1] - locs[0] + 1;
        loop:
        while (left < right) {
            mid = left + right >> 1;
            cnt = c;
            prev = locs[0];
            for (i = 1; i < n; i++) {
                if (locs[i] - prev >= mid) {
                    if (--cnt == 0) {
                        left = mid + 1;
                        continue loop;
                    }
                    prev = locs[i];
                }
            }
            right = mid;
        }
        System.out.print(right - 1);
    }
}
